### ISSUE TITLE
Memory Leak in `distance_transform_3d` Function Leading to OOM During Training

### ISSUE THREAD

**Issue by BubbleDirk** [kornia/kornia#3099]
### Describe the bug

During a pytorch-based CNN training, when using the `distance_transform` function, I observed a continuous increase in GPU memory consumption. After a few iterations, this leads to an Out of Memory (OOM) error. It should be noted that the GPU memory is not being released after each iteration, consequently leading to overflow after a few iterations.

### Reproduction steps

```bash
1.  Integrate the provided `distance_transform` function into a training loop.
2.  Monitor GPU memory usage (e.g., using `nvidia-smi`).
3.  Observe a steady increase in memory consumption with each iteration, eventually leading to an OOM error.
```

### Expected behavior

The `distance_transform` function should not cause a continuous increase in GPU memory usage during training. Memory used within the function should be released after each call.

### Environment

```shell
- PyTorch Version 1.8.1
- OS Ubuntu 20.04
- Python version 3.8.10
- CUDA/cuDNN version: 11.1
```

**Comment by edgarriba** [comment]
Do you have any idea what operator would do that and how to improve this ?

**Comment by BubbleDirk** [comment]
To address the OOM error during training, I implemented a preprocessing step where the input data was downsampled before being fed into the `distance_transform` function. Subsequently, the results were upsampled to restore the original dimensions. This approach effectively prevented the OOM errors that typically occurred around the 20th iteration when using the full-sized input. Logically, even with a for loop in the function, temporary variables should be reclaimed by the garbage collector after gradient backpropagation. If there's a problem with the garbage collector, it shouldn't be directly linked to the input data size. This inconsistency is puzzling to me.

**Comment by edgarriba** [comment]
feel free to re-open of there's still any interest in this. Thanks !
